### DISCUSSION TITLE
Mailboxes and forwarding

### DISCUSSION THREAD

**Discussion by simonwiles** [docker-mailserver/docker-mailserver#3523] [Q&A] [answered]
I've been reading through #3270 and the related issues, and I have a question, if I may.

I want to run a mail server where addresses have mailboxes attached (so that they can be logged in to, and so that mail is preserved on the server -- sending is not so important), but the mail must also be forwarded to one or more external addresses.  Do I understand this is no longer possible?

(I had this setup working with a previous version of DMS, but I see that it has been blocked now as unsupported?)

Does anyone know if there's a way to achieved this?  Very grateful for any suggestions!

Thanks :)

**Comment by polarathene** [comment]
> I want to run a mail server where addresses have mailboxes attached (so that they can be logged in to, and so that mail is preserved on the server -- sending is not so important)

Use `setup email add hello@example.com` to create a login account with mail storage.

> but the mail must also be forwarded to one or more external addresses

Use an alias to direct mail to the local storage and external addresses by assigning them both to the same alias:
- `setup alias add alias@example.com hello@example.com`
- `setup alias add alias@example.com world@external.com`

You may have [trouble with current `v12.1` release](https://github.com/docker-mailserver/docker-mailserver/issues/3358#issuecomment-1558166993) and need to use `:edge` until `v13` is released (_but this is a major version introducing breaking changes while it's worked on_).

Is this not the functionality you're asking for?
- Login via `hello@example.com` account (_eg: via Thunderbird / Roundcube to read mail via IMAP_)
- Receive mail to `alias@example.com` and it will deliver a copy to both recipient mailboxes: `hello@example.com` and `world@external.com`.
- Receive mail to `hello@example.com` directly would also be an option, but would not forward a copy to `world@external.com`.

That should still all be supported and fairly explicit.

---

Were you wanting to configure this differently, such as `hello@example.com` not being accepted directly, or not having the distinct separation of alias and account? What would that look like? (_[here are two alternative approaches](https://stackoverflow.com/a/33895835) outside of DMS_)

[`always_bcc`](https://www.postfix.org/postconf.5.html#always_bcc), [`recipient_bcc_maps`](http://www.postfix.org/postconf.5.html#recipient_bcc_maps), and [`sender_recipient_maps`](http://www.postfix.org/postconf.5.html#sender_bcc_maps) may meet your needs (_no official support in DMS, but you can add these easily with a [`postfix-main.cf` override](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/advanced/override-defaults/postfix/)_).
- The first allows you to set an address to send a copy of every email received by DMS to that chosen mail address (external or local).
- While the other two use a common lookup format with a pattern to match a Sender or Recipient address for mail arriving at DMS, and if the lookup config file has a match, an address to send a copy to is returned. This way you can send to a specific address depending on where mail is coming from or going.

---

There is also [this project](https://github.com/aspettl/mailadmin) may also be of interest, I haven't looked into it much but it may be another option that is said to have compatibility with DMS.

**Comment by simonwiles** [comment]
Thank you for these suggestions, I really appreciate your help 🙂

What I was really hoping was to be able to have, e.g., `treasurer@example.org` (`secretary@example.org` etc.) be a mail account which can be accessed via IMAP, mail preserved on the server and able to be backed up, etc., but also have mail that arrives there forwarded to `address-of-treasurer@institution.edu`.  I seemed to have this working with the DMS v11.2 that I was experimenting/prototyping with at the back-end of 2022.  The ability to send mail from `treasurer@example.org` was also possible, though that's something we can probably live without.

Does this seem like an unusual use-case?  To have all mail that arrives in an inbox forwarded to an external address?

I'm loathe to try to persuade a bunch of (mostly older) academics across four continents to change the way they're used to working (famously easy to achieve 🤣!), so I'll look into `recipient_bcc_maps`, which I think looks like my best bet.  I'll also take a look at https://github.com/aspettl/mailadmin since the README seems to suggest it might offer what we need -- perhaps I can learn something from what's going on there.

**Reply by polarathene** [reply]
> What I was really hoping was to be able to have, e.g., `treasurer@example.org` (`secretary@example.org` etc.) be a mail account which can be accessed via IMAP, mail preserved on the server and able to be backed up, etc.

- `setup email add ...` does this. It creates an account with a login name (email address) and password to authenticate with (_Postfix for sending, Dovecot for reading received mail_).
- Mail received to an account has it's storage handled by Dovecot at `/var/mail` in the container (_or your volume mount for persistence_). You can back that up from your host however you like, they're just files AFAIK so that should go smoothly.

> but also have mail that arrives there forwarded to `address-of-treasurer@institution.edu`.

`setup alias add ...` as noted would support this, but it seems your main concern was avoiding a separate mail address that this now enforces. Since you always want the mail forwarded, our alias support won't meet your requirement if your actual mail account receives mail directly...

We won't change the alias support, as this is functioning correctly. It's not for forwarding mail but redirecting it.

> The ability to send mail from `treasurer@example.org` was also possible, though that's something we can probably live without.

Any mail account should be able to login to send mail.

Aliases can also do this indirectly IIRC (although it resolves to the aliased account), provided they're a 1:1 mapping, not to multiple recipients. Alias login is not intentional, and is actually due a workaround for Dovecot Quotas support that is difficult to fix properly, but enabled by default.

The correct behaviour is `setup email add ...` configures accounts you can login with. Without `SPOOF_PROTECTION=1`, you can send mail as any sender address you like (_even if you don't own it, but expect the destination to probably reject it_).

> Does this seem like an unusual use-case? To have all mail that arrives in an inbox forwarded to an external address?

Not at all.

As you've referenced an issue where it was problematic in the past. I think it's more of a misunderstanding between users and maintainers. You've been helpful in clarifying the use-case when mail doesn't need to be redirected/relayed, but delivering locally while additionally forwarding a copy elsewhere (_where aliases are not sufficient since they are not the only route to a users mailbox_).

---

## Summary of Scenarios

**Accounts** are for login:
- They **represent an identity** with an associated mail address and storage (_both technically optional, with `SMTP_ONLY=1`?_).
- These accounts provide the **ability to authenticate as the mail account and send mail** as any sender address (_Unless DMS is configured to be more strict_).
- In DMS **we configure the username login as the mail address** created and associated to the account identity, although this **may not be required** in practice?

**Aliases** redirect mail, distinct from an account they are not intended for login (_eg: What account would `marketing@example.com` login into below?_):
- Alias for forward-only (_no local delivery storage in DMS_) to an external mail service.
- Alias to deliver mail to one or more actual mail accounts (_`marketing@example.com` => `john@example.com` , `jane@example.com`, `bob@not-example.com`_).
- An alias could **indirectly support delivery to a DMS mail account and forward a copy to another address** (_but as it is only redirection without mail storage itself, mail may opt-out by delivering directly to the DMS mail account address instead **where no forwarding is performed**_). The alias does not necessarily represent the only inbound address for an account, there could also be multiple aliases where forwarding is not desirable, in that sense the mail accounts owned address is not unlike a separate alias. This scenario / expectation for forwarding is a misunderstanding of aliases?

### Known issues with use-cases trying to leverage alias feature
- Users want **an alias to forward mail to an external provider** (_like Gmail for storage and use the WebUI / phone app_), **no storage in DMS**. But they want **mail to be sent from DMS** (relay host).
  - An **alias is not intended for login/authentication**, so while it can forward the mail to Gmail, to send mail from DMS via relay host (_or directly via a MUA like ThunderBird_) **would require login via a mail account**.
  - **Users may get confused with this overlap** when thinking of a common mail address to receive (_to receive by DMS and forward_) and send (_by needing to login via DMS account_), but trip up on the alias vs account addresses not allowing overlap.
- **Deliver to local storage in DMS**, but **always forward a copy elsewhere** too (internally or externally).
  - Alias as explained earlier **can't quite support this** when it's not the only public facing address to an account. The fact that it **doesn't support a single reachable address** (_due to needing a mail account in DMS to support local delivery_) is also an issue.
  - Multiple aliases to an account would also be tedious to configure this way for each recipient(s) needing to have another copy forwarded (_potentially selectively_).
  - **Requires an alternative approach** perhaps with BCC settings in Postfix `main.cf` to properly configure a mail processed through Postfix to always be forwarded (_eg: for all mail received, or based on sender/recipient address such as a specific DMS mail account as a recipient_).

---

Perhaps there is another way to approach the use-case for forwarding. I think it's ok to raise a feature request for this and referencing this comment. Someone would need to implement it though.

**Comment by simonwiles** [comment]
Update: I think I have `recipient_bcc_maps` working how I want it to, thank you! 🎉

It's a little complicated and I need an arbitrary intermediary alias to be able to forward to more than one address, but I *think* it's working.  The outbound BCC mail is not being signed with DKIM, but I'll start working on that next, I guess.

**Reply by polarathene** [reply]
> It's a little complicated and I need an arbitrary intermediary alias to be able to forward to more than one address,

Ah I see, that is limited to only a single address for BCC? I misunderstood and thought the sender / recipient lookup tables supported mapping to multiple addresses.

The alias workaround you're using for that isn't ideal and creates an inverse of the aliases with accounts concern raised.

You state that you need to forward to more than one address? (_which is what an alias is good for_) I'm a little confused as your use-case scenario was for a DMS mail account to forward a copy to an equivalent external address for that account? But now you also want it to forward to additional addresses too?

There is `.forward` file which works like our virtual alias support, but is only for `local` delivery agent (_your account uses the `virtual` delivery agent which directs mail received to Dovecot, while Postfix has it's own `local` alternative if not using Dovecot_). Doesn't seem like we can easily support that.

I will give this some more thought 😅 

---

> The outbound BCC mail is not being signed with DKIM

Might depend on OpenDKIM vs Rspamd DKIM support. @georglauterbach may have input here.

A quick guess might be that it's presently tied to mail submission ports (465 / 587) and not port 25.

I would need to refresh on DKIM, but I think that technically DMS only needs one DKIM private key to sign for any outbound mail, but could optionally be configured for individual mail domains to lookup different private keys, whereas the mail domains DNS provides a DNS record for the public key to verify signatures with (_which could be that single or individual private key setup, but DMS may only be taking an individual domain lookup approach?_).

Technically DMS could send mail out for a sender address it's authorized to send as, without being the MX host that mail for that address is delivered to, thus no mail account for that address configured... which may affect the assumptions for how our current support approached implementing DKIM signatures.

Alternatively, maybe related to the submission ports guess, it may only do so for authenticated mail submission. Thus relaying/forwarding from DMS when mail is received, especially when outbound port is 25 may not apply a DKIM signature.

I do know that there was some DKIM issues resolved in `v13` (unreleased), so you might have luck with `:edge` if you want to try test that.

**Reply by georglauterbach** [reply]
In case you're using Rspamd, let me know if you encounter any issues; I'm maintaining the implementation. Please open an issuen though as I am low on time and cannot read through discussions really.

**Comment by polarathene** [comment]
## Known use-cases

### Forward-only

**User clarifies that** (_in below "References"_):
- They want mail to forward to `@other.com` accounts (Gmail) and are **not interested in local mail storage** (_or Gmail logging into Dovecot in DMS to retrieve mail_).
- They want to **send mail via DMS from Gmail**, acknowledging that a DMS mail account is required for authenticated submission.
- **They did not realize** that the mail account did not need to match the address they want to send as, **which caused the conflict with aliases**.

In this example, the users config AFAIK would resolve `group-1` to two addresses, one being `test1@domain.com` which is ambiguous from the earlier commands.

```bash
# DMS mail accounts for storage and sending as:
setup email add test1@domain.com
setup email add test2@domain.com

# Alias created with same mail account address, to forward mail to alternative external address:
setup alias add test1@domain.com test1@other.com
setup alias add test2@domain.com test2@other.com

# Alias created for groups that send mail to DMS mail accounts / aliases:
setup alias add group-1@domain.com test1@domain.com
setup alias add group-1@domain.com test2@domain.com
setup alias add group-2@domain.com test1@domain.com
setup alias add group-2@domain.com test2@domain.com
```

Does Postfix alias `test1@domain.com` recipient of alias `group-1@domain.com` (_receiving a mail at DMS_):
- To the external forward address `test1@other.com`?
- Recognize the local account it could deliver mail to?
- Acknowledge and send to both?

Postfix will process virtual aliasing (_`test1@domain.com` => test1@other.com`_) before it checks virtual mailboxes (_local DMS account `test1@domain.com`_), so technically AFAIK it should only send to `test1@other.com`.

If the user did want mail to reach the local DMS mail storage, it might work if an alias was configured to point to itself as a recipient too. [Postfix docs for the `virtual` delivery agent](https://www.postfix.org/virtual.5.html) mention that aliasing is recursive and that it will terminate **when that address aliases to itself**:

> _Virtual aliasing is recursive; to terminate recursion for a specific address, alias that address to itself._

<details>
<summary>References</summary>

https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1508580393

> _I've my own domain which is supported by your product but I'm forwarding emails to Google for easier management - that's why I have aliases to `other.com`._
> _But to send emails from from google as my domain I need email account as well - right???_
> 
> _Aliases for groups are to support emails from 3rd party services that suppose to be forwarded to different recipients._

https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1508703199

> - _I do not want to send email using Google as relay_
> - _I would like to send emails and use my server as relay with auth - looks like I need local email accounts to achieve that_

https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1511105995

> _My setup looks like that:_
> 1. _I have my own DMS_
> 2. _I'm forwarding all messages from accounts to gmail_
> 3. _I'm sending emails from gmail as user from DMS with authentication_
> 4. _I'm not able to add DMS account to more then 1 group alias_

https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1511261934

> _I do not want gmail to login to the server to get emails_

</details>

The [suggested solution](https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1509366419) at the time was to:
- Have Gmail pull mail that was delivered to DMS mail storage via POP3 (_to remove from DMS_).
- Send through DMS via that same authenticated mail account at DMS (_that is associated to that mail storage_).
- [An alternative suggestion](https://github.com/docker-mailserver/docker-mailserver/pull/3259#issuecomment-1513294364) (_misunderstanding on my part_) was to configure `relay_domains`.

**Users desired setup was viable** if they used a mail account with different address for login to send mail. Their Gmail setup seems like it wasn't needing to scale for a large number of users, so that sounds perfectly fine to **use a single login account with DMS** to send mail through (_as any of the aliases_).

---

<details>
<summary>3 similar reports (forwarding to gmail, no storage on DMS)</summary>

**2nd report:** https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1547006955

> _My use-case is to have a catch-all address (i.e. `@domain.com` that redirects to an external service)._
> - _There are some addresses where I have overridden the aliases and they have a matching account (i.e. `user1@domain.com` exists both as an alias and as an email account) in order to be able to send mail as well._
> - _No email is stored on DMS._
> - _This had been working flawlessly for the last few years._

```bash
# Create catch-all alias, all mail sent to `@domain.com` goes to `my.account@other.com`:
setup alias add @domain.com my.account@other.com

# Add override, mail sent to this address should go to `user1@other.com` instead:
setup alias add user1@domain.com user1@other.com

# Added mail accounts for authenticated submission to send mail:
setup account add me@domain.com
setup account add user1@domain.com
```

Same scenario mostly:
- Forward mail (_via aliases_) delivered to DMS to an external mail address.
- Added mail accounts for sending, but from `v12.1` of DMS became **unsupported to have both an alias and account with the same address** assigned.

Confusion was cleared up with a [follow-up response and advice](https://github.com/docker-mailserver/docker-mailserver/issues/3262#issuecomment-1547041446) from me to use a separate login to send mail with, or a potential workaround to bypass our prevention.

---

**3rd report:** https://github.com/docker-mailserver/docker-mailserver/issues/3350

Basically same as the 2nd report, user is unable to add an account and alias with same address, and only describes wanting to forward mail to Gmail.

[I request more information on use-cases for configuring](https://github.com/docker-mailserver/docker-mailserver/issues/3350#issuecomment-1550528898) to better understand the problem (_no response received_). I also listed a variety of other setup choices that might work for the user. I identified their setup as using DMS as a relay (_mail arrives to DMS to be sent out to another address_).

Although in this case no authentication is necessary, if DMS accepts the mail, it's forwarding it on behalf of the user via the alias, which is a bit different than relay host with authenticated submission or sending mail on behalf of a sender address DMS doesn't manage.

---

**4th report:** https://github.com/docker-mailserver/docker-mailserver/issues/3437

Another user is trying to follow our outdated community contributed guide, sending mail arriving at DMS address to forward to a gmail address.

A [decently formatted and detailed response from me](https://github.com/docker-mailserver/docker-mailserver/issues/3437#issuecomment-1646350088) takes another shot at this use-case but receives no response again.

In this use-case, they probably wanted to use authenticated submission to send directly to the gmail account rather than a forwarding alias as this was for a local process on the host to send a notification, which may not require a public address that anyone could send mail to forward to their gmail address.

</details>

---

The original 1st report detailed earlier above had an **error with the "group" alias** to multiple recipients, but this only happened when adding the recipient that an existing alias referenced.

**This was their actual failure** with the `setup alias add` command output, and was [reported and tracked down here](https://github.com/docker-mailserver/docker-mailserver/issues/3358#issuecomment-1556918175) and [resolved](https://github.com/docker-mailserver/docker-mailserver/pull/3359) for v13.

In the meantime until v13 is out, a suggestion was to run the `setup` command via the `:edge` image, but otherwise continue running DMS via the `:v12.1` image tag.

---

### Delivery to DMS account + forwarding a copy

Your [issue discussed here](https://github.com/orgs/docker-mailserver/discussions/3523#discussioncomment-6910225) is the only one I'm aware of that is wanting to also keep the original mail delivered to DMS stored with Dovecot, in addition to the forwarded copy.

Technically, if the BCC feature doesn't meet your needs. Having a virtual alias resolving to itself (_and a DMS account with the same address_) and the additional forwarded address(es) should work. DMS `setup alias add` will prevent this though since `v12.1`.

I'm not sure that aliases self-referencing themselves to support forwarding with local delivery to an address of the same alias is intended to be used that way though? But if more users require it we could consider supporting it. I'd like it if we can get another viable way to support that use-case, or at the very least confirmation that some other big mail project (_mailu, mailcow, etc_) is allowing and using it for forwarding support like this.

**Reply by polarathene** [reply]
Feature request raised if you'd like to subscribe to it and 👍 react. I don't expect it'll go anywhere but who knows 😅

**Reply by simonwiles** [reply]
Thank you -- I'll take a close look at the FR.

> Having a virtual alias resolving to itself (and a DMS account with the same address) and the additional forwarded address(es) should work. DMS setup alias add will prevent this though since v12.1.

That's what was working for me previously, I think -- seemed intuitive and effective.  But yes, upgrading to v12.1 I saw this was now blocked.  I don't really understand the reasoning behind #3270, but if using `recipients_bcc_maps` to achieve the same thing is better/preferred for some reason then I'll do that (albeit it's more complicated and seems less intuitive -- to me, at least).

**Reply by polarathene** [reply]
I think the main motivation behind preventing it was that it did not make sense to have:

- Mail account with login => `hello@example.com`
- Mail alias `hello@example.com` => `hello@gmail.com` (_now mail cannot go to `hello@example.com` storage in DMS, it just gets redirected to Gmail instead_)

There was 4 times where users reported this exact setup and it made no sense. They only created an account in DMS with the same address to send mail, but that wasn't required to actually share the same mail address as the alias.

We figured preventing that was the right approach as no other use-case justified it. Yours does and with some more consideration we could probably revert the active prevention. Hence why I documented the known use-cases above and raised an FR about it. It's possible that the maintainers made a mistake here (_we thought of it as a bug fix_).

---

> f using `recipients_bcc_maps` to achieve the same thing is better/preferred for some reason then I'll do that (albeit it's more complicated and seems less intuitive -- to me, at least).

Is that partly because you're manually configuring it?

You can technically do the exact same thing with the `virtual_alias_maps` setting. It uses a lookup table(s) in a similar format (_alias recipient pattern, followed by one or more recipients to resolve to_). DMS only actively prevents with `setup alias add` / `setup email add`. I think you can still leverage those for everything else, and just manually add the alias into `postfix-virtual.cf` (_or even run the `setup alias add ...` with the `v12.0` image tag temporarily_).

---

Unrelated to the prevention change, but another common alias issue we have seen:

There are some user reports where they used the command the wrong way, by making the alias `setup alias add hello@gmail.com hello@example.com` which would redirect mail sent to DMS for `hello@gmail.com`, **and** when sending mail from DMS to `hello@gmail.com` directly, DMS would think "hey I own this mail alias, I should handle it" and it would go into `hello@example.com` mailbox, not delivered to gmail. I think the worse outcome was that any mail sent to a `@gmail` account would also try to route back into DMS but fail to find a lookup match and reject it, failing to deliver.

**Reply by simonwiles** [reply, marked as answer]
Okay, I think I understand -- I can do it the way I was in DMS v11.2, I just need to add the entries to `postfix-virtual.cf` by hand; `setup.sh` refusing to do so and the related PR referring to this as unsupported are not as serious as I thought -- it just means the DMS project has found that workflow to mislead some folks, so prefers not to promote it (besides, my `recipients_bcc_maps` solution is just as "unsupported" by that same measure).  I think I'll revert the `recipients_bcc_maps` solution -- it was an interesting and instructive episode, though -- thanks! 🙂

edit: to answer your question directly

> Is that partly because you're manually configuring it?

Yeah adding the `recipients_bcc_maps` meant I had to use an intermediate alias anyway (in the fairly common case where I want to forward mail for one address to two or more external recipients, at least), and I had to use an extra file, and use `postmap` to compile the map (until I learned about `texthash`) and all-in-all it feels like I'll need more internal documentation to remind myself and others how it works for the future 🙂

**Reply by polarathene** [reply]
> it just means the DMS project has found that workflow to mislead some folks, so prefers not to promote it

Yep, but we also weren't aware of a valid use-case that you've shared, so that's also a misunderstanding on our end (_if there isn't a better alternative to support it_).

We can probably address the main concern with better communication in our documentation, rather than actively prevent it if it's not introducing any other problems with DMS.

> it feels like I'll need more internal documentation to remind myself and others how it works for the future 🙂

Yeah, not fun :(

Keep an eye on the changelog / release notes, and eventually this might be reverted.

> it was an interesting and instructive episode, though -- thanks! 🙂

Cheers for being so responsive! ❤️
